core/services/head_relayer.go

```diff
@@ -0,0 +1,146 @@
+package services
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+const callbackTimeout = 2 * time.Second
+
+type callbackID [256]byte
+
+// HeadRelayable defines the interface for listeners
+type HeadRelayable interface {
+	OnNewLongestChain(ctx context.Context, head models.Head)
+}
+
+// NewHeadRelayer creates a new HeadRelayer
+func NewHeadRelayer() *HeadRelayer {
+	return &HeadRelayer{
+		callbacks:     make(map[callbackID]HeadRelayable),
+		mailbox:       utils.NewMailbox(1),
+		mutex:         &sync.RWMutex{},
+		chClose:       make(chan struct{}),
+		wgDone:        sync.WaitGroup{},
+		StartStopOnce: utils.StartStopOnce{},
+	}
+}
+
+// HeadRelayer relays heads from the head tracker to subscribed jobs, it is less robust against
+// congestgion than the head tracker, and missed heads should be expected by consuming jobs
+type HeadRelayer struct {
+	callbacks map[callbackID]HeadRelayable
+	mailbox   *utils.Mailbox
+	mutex     *sync.RWMutex
+	chClose   chan struct{}
+	wgDone    sync.WaitGroup
+	utils.StartStopOnce
+}
+
+var _ store.HeadTrackable = (*HeadRelayer)(nil)
+
+func (hr *HeadRelayer) Start() error {
+	return hr.StartOnce("HeadRelayer", func() error {
+		go hr.run()
+		return nil
+	})
+}
+
+func (hr *HeadRelayer) Close() error {
+	if !hr.OkayToStop() {
+		return errors.New("HeadRelayer is already stopped")
+	}
+	close(hr.chClose)
+	hr.wgDone.Wait()
+	return nil
+}
+
+func (hr *HeadRelayer) Connect(head *models.Head) error {
+	return nil
+}
+
+func (hr *HeadRelayer) Disconnect() {}
+
+func (hr *HeadRelayer) OnNewLongestChain(ctx context.Context, head models.Head) {
+	hr.mailbox.Deliver(head)
+}
+
+func (hr *HeadRelayer) Subscribe(callback HeadRelayable) (unsubscribe func()) {
+	hr.mutex.Lock()
+	defer hr.mutex.Unlock()
+	id, err := newID()
+	if err != nil {
+		logger.Errorf("Unable to create ID for head relayble callback: %v", err)
+		return
+	}
+	hr.callbacks[id] = callback
+	return func() {
+		hr.mutex.Lock()
+		defer hr.mutex.Unlock()
+		delete(hr.callbacks, id)
+	}
+}
+
+func (hr *HeadRelayer) run() {
+	hr.wgDone.Add(1)
+	defer hr.wgDone.Done()
+	for {
+		select {
+		case <-hr.chClose:
+			return
+		case <-hr.mailbox.Notify():
+			hr.executeCallbacks()
+		}
+	}
+}
+
+// DEV: the head relayer makes no promises about head delivery! Subscribing
+// Jobs should expect to the relayer to skip heads if there is a large number of listeners
+// and all callbacks cannot be completed in the allotted time.
+func (hr *HeadRelayer) executeCallbacks() {
+	hr.mutex.RLock()
+	defer hr.mutex.RUnlock()
```

Comment:
This keeps the mutex locked for a long time. Maybe we could lock just long enough to copy the slice and then unlock and use our copied version?